File: Aigilas/Aigilas/Strategies/ControlledByPlayer.cs
using System;
using SPX.Entities;
using Aigilas.Dungeons;
using Aigilas.Entities;
using Aigilas.Management;
using Aigilas.Creatures;
using SPX.Core;
using System.Collections.Generic;
namespace Aigilas.Strategies
{
    public class ControlledByPlayer : IStrategy
    {
        private bool _isCasting = false;
        private Point2 _keyVelocity = new Point2(0, 0);

        private static readonly List<int> __hotkeys = new List<int>()
        {
            Commands.HotSkill1,
            Commands.HotSkill2,
            Commands.HotSkill3
        };

        public ControlledByPlayer(ICreature parent) : base(parent,Strategy.ControlledByPlayer)
        {
            _targets.AddTargetTypes(ActorType.NONPLAYER);
        }

        public override void Act()
        {
            if (Input.IsActive(Commands.Start, _parent.GetPlayerIndex()))
            {
                _parent.SetPlaying(true);
            }
            if (Input.IsActive(Commands.Back, _parent.GetPlayerIndex()))
            {
                _parent.SetPlaying(false);
            }
            if (_parent.IsPlaying())
            {
                if (!Input.IsContext(Contexts.Inventory,_parent.GetPlayerIndex()))
                {
                    var leftVelocity = (Input.IsActive(Commands.MoveLeft, _parent.GetPlayerIndex()) ? -Stats.DefaultMoveSpeed : 0);
                    var rightVelocity = ((Input.IsActive(Commands.MoveRight, _parent.GetPlayerIndex())) ? Stats.DefaultMoveSpeed : 0);
                    _keyVelocity.SetX(rightVelocity + leftVelocity);

                    var downVelocity = ((Input.IsActive(Commands.MoveDown, _parent.GetPlayerIndex())) ? Stats.DefaultMoveSpeed : 0);
                    var upVelocity = ((Input.IsActive(Commands.MoveUp, _parent.GetPlayerIndex())) ? -Stats.DefaultMoveSpeed : 0);
                    _keyVelocity.SetY(upVelocity + downVelocity);

                    if (Input.IsContext(Contexts.Free, _parent.GetPlayerIndex()))
                    {
                        var isPress = Input.IsActive(Commands.Confirm, _parent.GetPlayerIndex());
                        if (!isPress)
                        {
                            _parent.SetInteraction(false);
                        }
                        if (isPress && !_parent.IsInteracting())
                        {
                            _parent.SetInteraction(true);
                        }
                        var skillCycleVelocity =
                            ((Input.IsActive(Commands.CycleLeft, _parent.GetPlayerIndex())) ? -1 : 0)
                            +
                            ((Input.IsActive(Commands.CycleRight, _parent.GetPlayerIndex())) ? 1 : 0);
                        _parent.CycleActiveSkill(skillCycleVelocity);

                        if (!_isCasting)
                        {
                            if (!Input.IsActive(Commands.Confirm, _parent.GetPlayerIndex(),false))
                            {
                                _parent.MoveIfPossible(_keyVelocity.X, _keyVelocity.Y);
                            }
                            if (!_keyVelocity.IsZero())
                            {
                                _parent.SetSkillVector(_keyVelocity);
                            }
                        }
                    }
                    if (Input.IsActive(Commands.Skill, _parent.GetPlayerIndex()))
                    {
                        _isCasting = true;
                    }

                    foreach (var hotkey in __hotkeys)
                    {
                        if (Input.IsActive(hotkey,_parent.GetPlayerIndex()))
                        {
                            if (!Input.IsActive(Commands.LockSkill, _parent.GetPlayerIndex(), false))
                            {
                                if (_parent.SetHotSkillActive(hotkey))
                                {
                                    _isCasting = true;
                                }
                            }
                            else
                            {
                                _parent.MarkHotSkill(hotkey);
                            }
                        }
                    }

                    if (_isCasting)
                    {
                        if (_parent.GetSkillVector() == null)
                        {
                            _parent.SetSkillVector(new Point2(1, 0));
                        }
                        if (!_parent.GetSkillVector().IsZero())
                        {
                            _parent.UseActiveSkill();
                            _isCasting = false;
                        }
                    }
                }
                
                if (Input.IsActive(Commands.Inventory, _parent.GetPlayerIndex()))
                {
                    Input.SetContext(_parent.ToggleInventoryVisibility()? Contexts.Inventory: Contexts.Free, _parent.GetPlayerIndex());
                }
            }
        }
    }
}

File: aigilas/src/aigilas/strategies/impl/ControlledByPlayer.java
package aigilas.strategies.impl;

import aigilas.Aigilas;
import aigilas.AigilasConfig;
import aigilas.creatures.BaseCreature;
import aigilas.creatures.Stats;
import aigilas.creatures.impl.Player;
import aigilas.strategies.BaseStrategy;
import aigilas.strategies.Strategy;
import sps.bridge.*;
import sps.core.Point2;
import sps.entities.Entity;
import sps.entities.EntityManager;
import sps.io.Input;

import java.util.Arrays;
import java.util.List;

public class ControlledByPlayer extends BaseStrategy {
    private boolean _isCasting = false;
    private final Point2 _keyVelocity = new Point2(0, 0);

    private static final List<Command> __hotkeys = Arrays.asList(Commands.get(Aigilas.Commands.Hot_Skill_1), Commands.get(Aigilas.Commands.Hot_Skill_2), Commands.get(Aigilas.Commands.Hot_Skill_3));

    public ControlledByPlayer(BaseCreature parent) {
        super(parent, Strategy.ControlledByPlayer);

        _targets.addTargetTypes(new sps.bridge.ActorType[]{ActorTypes.get(Sps.ActorGroups.Non_Player)});
    }

    @Override
    public void act() {
        if (Input.get().isActive(Commands.get(Aigilas.Commands.Start), _parent.getPlayerIndex())) {
            _parent.setPlaying(true);
        }
        if (Input.get().isActive(Commands.get(Aigilas.Commands.Back), _parent.getPlayerIndex())) {
            _parent.setPlaying(false);
        }
        if (_parent.isPlaying()) {
            if (!Input.get().isContext(Contexts.get(Aigilas.Commands.Inventory), _parent.getPlayerIndex())) {
                float leftVelocity = (Input.get().isActive(Commands.get(Aigilas.Commands.MoveLeft), _parent.getPlayerIndex()) ? -Stats.DefaultMoveDistance : 0);
                float rightVelocity = ((Input.get().isActive(Commands.get(Aigilas.Commands.MoveRight), _parent.getPlayerIndex())) ? Stats.DefaultMoveDistance : 0);
                _keyVelocity.setX(rightVelocity + leftVelocity);
                if (_keyVelocity.X > 0) {
                    _parent.setFacingLeft(false);
                }
                if (_keyVelocity.X < 0) {
                    _parent.setFacingLeft(true);
                }
                float downVelocity = ((Input.get().isActive(Commands.get(Aigilas.Commands.MoveDown), _parent.getPlayerIndex())) ? -Stats.DefaultMoveDistance : 0);
                float upVelocity = ((Input.get().isActive(Commands.get(Aigilas.Commands.MoveUp), _parent.getPlayerIndex())) ? Stats.DefaultMoveDistance : 0);
                _keyVelocity.setY(upVelocity + downVelocity);

                if (Input.get().isContext(Contexts.get(Sps.Contexts.Free), _parent.getPlayerIndex())) {
                    boolean isPress = Input.get().isActive(Commands.get(Aigilas.Commands.Confirm), _parent.getPlayerIndex());
                    if (!isPress) {
                        _parent.setInteraction(false);
                    }
                    if (isPress && !_parent.isInteracting()) {
                        _parent.setInteraction(true);
                    }
                    int skillCycleVelocity = ((Input.get().isActive(Commands.get(Aigilas.Commands.CycleLeft), _parent.getPlayerIndex())) ? -1 : 0) + ((Input.get().isActive(Commands.get(Aigilas.Commands.CycleRight), _parent.getPlayerIndex())) ? 1 : 0);
                    _parent.cycleActiveSkill(skillCycleVelocity);

                    if (!_isCasting) {
                        if (!Input.get().isActive(Commands.get(Aigilas.Commands.Confirm), _parent.getPlayerIndex(), false)) {
                            _parent.moveIfPossible(_keyVelocity.X, _keyVelocity.Y);
                        }
                        if (!_keyVelocity.isZero()) {
                            _parent.setSkillVector(_keyVelocity);
                        }
                    }
                }
                if (Input.get().isActive(Commands.get(Aigilas.Commands.Skill), _parent.getPlayerIndex())) {
                    _isCasting = true;
                }

                for (Command hotkey : __hotkeys) {
                    if (Input.get().isActive(hotkey, _parent.getPlayerIndex())) {
                        if (!Input.get().isActive(Commands.get(Aigilas.Commands.LockSkill), _parent.getPlayerIndex(), false)) {
                            if (_parent.setHotSkillActive(hotkey)) {
                                _isCasting = true;
                            }
                        }
                        else {
                            _parent.markHotSkill(hotkey);
                        }
                    }
                }

                if (_isCasting) {
                    if (_parent.getSkillVector() == null) {
                        _parent.setSkillVector(new Point2(1, 0));
                    }
                    if (!_parent.getSkillVector().isZero()) {
                        _parent.useActiveSkill();
                    }
                    _isCasting = false;
                }
            }

            if (Input.get().isActive(Commands.get(Aigilas.Commands.Inventory), _parent.getPlayerIndex())) {
                if (AigilasConfig.get().debugInventory) {
                    for (Entity player : EntityManager.get().getPlayers()) {
                        Player p = (Player) player;
                        if (p.getPlayerIndex() != _parent.getPlayerIndex()) {
                            p.toggleInventoryVisibility();
                        }
                    }
                }
                Input.get().setContext(_parent.toggleInventoryVisibility() ? Contexts.get(Aigilas.Commands.Inventory) : Contexts.get(Sps.Contexts.Free), _parent.getPlayerIndex());
            }
        }
    }
}
